File: 프로그래머스/2/84512. 모음 사전/모음 사전.js
const set = new Set();
const alp = ['A', 'E', 'I', 'O', 'U'];

function solution(word) {
    for (let i=0; i<alp.length; i++){
        makeLetter(alp[i], 1);
    }
    const allLetters = [];
    for (let str of set){
        allLetters.push(str);
    }
    allLetters.sort((a, b)=>(a+b)-(b+a));
    return allLetters.indexOf(word)+1;
}

function makeLetter(letter, idx){
    
    if (!set.has(letter)){
        set.add(letter);
    }
    if (idx===5){
        return;
    }
    for (let i=0; i<alp.length; i++){
        makeLetter(letter+alp[i], idx+1);
    }
}

File: 프로그래머스/2/84512. 모음 사전/모음 사전.java
import java.util.*;
class Solution {
    static List<String> madeWords = new ArrayList<>();
    static String[] vowels = new String[]{"A", "E", "I", "O", "U"};
    public int solution(String word) {
        int answer = 0;
        
        
        
        for (int i=0; i<vowels.length; i++){
            makeWord(vowels[i]);
        }
        
        Collections.sort(madeWords, new Comparator<String>(){
            @Override
            public int compare(String s1, String s2) {
                return s1.compareTo(s2);
            }
        });
        
        return madeWords.indexOf(word)+1;
    }
    public void makeWord(String word){
        if (!madeWords.contains(word)) madeWords.add(word);
        
        if (word.length()==5) return;
        for (int i=0; i<vowels.length; i++){
            makeWord(word+vowels[i]);
        }
    }
}